lxd/db/instances.go

```diff
@@ -338,35 +337,39 @@ func (c *Cluster) InstanceList(filter *InstanceFilter, instanceFunc func(inst In
 			return fmt.Errorf("Failed loading instances: %w", err)
 		}
 
+		profiles, err := cluster.GetProfiles(ctx, tx.tx, cluster.ProfileFilter{})
+		if err != nil {
+			return err
+		}
+
 		projects, err := cluster.GetProjects(ctx, tx.tx, cluster.ProjectFilter{})
 		if err != nil {
-			return fmt.Errorf("Failed loading projects: %w", err)
+			return err
 		}
 
-		// Index of all projects by name and record which projects have the profiles feature.
-		for _, project := range projects {
-			apiProject, err := project.ToAPI(ctx, tx.tx)
-			if err != nil {
-				return err
+		for _, instance := range instances {
+			instanceProfiles[instance.Name] = []api.Profile{}
+			for _, profile := range profiles {
+				if shared.StringInSlice(profile.Name, instance.Profiles) {
+					apiProfile, err := profile.ToAPI(ctx, tx.tx)
+					if err != nil {
+						return err
+					}
+
+					instanceProfiles[instance.Name] = append(instanceProfiles[instance.Name], *apiProfile)
```

Comment:
Multiple instances of the same name can exist in different projects, therefore unless you are in a single project context (which we aren't here), then it is never save to index an instance (or any of its properties) by instance name only, as you will get unexpected behavior when dealing with instances in different projects.

I preferred the original approach of getting all profiles and then indexing them by project and profile name.
That way we only load them and use `ToAPI` once for each profile (rather than for every instance), and we can then just add them to the instance from the index map as we iterate through each instance.

Reply:
Ah yeah, good catch. I did have an inkling that the refactor was too simple to be right. :p 

My main motivation for changing the logic was that I've noticed that as I've gone around refactoring the db package, I tend to spend a lot of time repeatedly working out exactly how everything interrelates in `InstanceList` each time I change something having to do with instances, profiles, or projects. There's quite a few maps and slices to keep track of :)

We definitely should add a test for the case I missed though.

Reply:
> We definitely should add a test for the case I missed though.

Test case sounds good, you can add it in the other PR you're going to do, thanks.